### PR TITLE
Active dev

### PR DESCRIPTION
**Pull request by slsevilla**
Ready to merge changes with main branch. Please review @skchronicles

### CODE REVIEW

File: aasap [hunk 1]
```diff
@@ -263,6 +363,248 @@ def run(sub_args):
                         "Please review the following mutated coding DNA sequence for any errors:\n\t> {}".format(mutated_dna))
 
 
+def predict(sub_args):
+    """Runs the prediction tool netMHC following output of the Amino Acid sequence analysis pipeline.
+    """
+    # set path for programs
+    netmhc_path="/data/CCBR_Pipeliner/bin/netMHC/netMHCpan-4.1/netMHCpan"
```
Comment (skchronicles) [inline aasap:370]: Remove the hard-coded path to `netMHCpan`. 
Let's assume that the tool is already in the user's path (update the docs to reflect this assumption). Please add some functionality to check to see `netMHCpan`. If it is not, then `fatal()` with a helpful message.

File: aasap [hunk 2]
```diff
@@ -263,6 +363,248 @@ def run(sub_args):
...
+
+    # Check kmer length is an odd number
+    if (sub_args.kmerLength % 2 ) == 0:
```
Comment (skchronicles) [inline aasap:373]: Replace with `fatal()`

File: aasap [hunk 3]
```diff
@@ -263,6 +363,248 @@ def run(sub_args):
...
+        print("WARNNING: --kmerLength must be an odd number. Please revise input and try again")
+        sys.exit()
+
+    # Check allele list is <=20 in length
+    split_alleleList=sub_args.alleleList.split(",")
+    if len(split_alleleList)>20:
+        print("WARNNING: --kmerLength must be an odd number. Please revise input and try again")
```
Comment (skchronicles) [inline aasap:380]: Replace with `fatal()`

File: aasap [hunk 4]
```diff
@@ -263,6 +363,248 @@ def run(sub_args):
...
+        sys.exit()
+
+    # Parse field of interest from each excel file. This file can be the output of AASAP run
+    # or maybe an input created by a user. Input file is required have the following fields:
+    # - Hugo_Symbol
+    # - Transcript_ID
+    # - Variant_Classification	
+    # - WT_Subset_AA_Sequence
+    # - Mutated_Subset_AA_Sequence
+    print("--Preparing Data")
```
Comment (skchronicles) [inline aasap:390]: Write logging information to standard error with `err()`

File: aasap [hunk 5]
```diff
@@ -263,6 +363,248 @@ def run(sub_args):
...
+    process = "{} -f {} -a {} -l {} -BA -xls -xlsfile {} > {}".format(
+        netmhc_path, 
+        netMHC_input,
+        sub_args.alleleList,
+        sub_args.peptideLength,
+        netmhc_raw_output,
+        netmhc_cmd_output
+    )
+
+    print("--Running: " + process)
```
Comment (skchronicles) [inline aasap:515]: Write logging information to standard error with `err()`

File: aasap [hunk 6]
```diff
@@ -263,6 +363,248 @@ def run(sub_args):
...
+    exitcode = bash(process)
+
+    # Read in output of netMHC
+    # Remove rows which includes the first row of only the allele name and 
+    # the second row which includes the header (header consists of dup values)
+    print("--Post-Processing")
```
Comment (skchronicles) [inline aasap:521]: Write logging information to standard error with `err()`

**Comment by skchronicles** [comment]
Also, please add `numpy` to the requirements.txt

**Comment by skchronicles** [comment]
At some point, the README will need to be updated with information pertaining to the `predict` and `input` subcommands.

**Comment by skchronicles** [comment]
@slsevilla Please let me know what you think.

File: aasap [hunk 7]
```diff
@@ -12,6 +12,9 @@ Example:
 """
 
 from __future__ import print_function
+from operator import sub
```
[the remark this answers is not in the document]
Comment (slsevilla) [reply aasap:15]: not needed. can remove.

File: aasap [hunk 8]
```diff
@@ -109,6 +114,100 @@ def build(sub_args):
...
 
 
+def input(sub_args):
+    """Creates input files for the Amin Acid Sequence Analysis Pipeline from MAF files.
+    Uses user-inputs to determine filtering thresholds.
+    """
+    # Check that col names exist, if they don't exit and error
+    def check(col,df_in):
+        if col not in df_in:
+            print ("The following column is required in input", col)
```
[the remark this answers is not in the document]
Comment (slsevilla) [reply aasap:124]: done

File: aasap [hunk 9]
```diff
@@ -109,6 +114,100 @@ def build(sub_args):
...
+            quit()
+    
+    # for each input file, read in file
+    # check for colun header and create variant id (VIDA)
+    # calculate the average VAF and append df to list
+    print("--Processing input files")
```
[the remark this answers is not in the document]
Comment (slsevilla) [reply aasap:130]: I wanted this to print to the terminal as a 'user update' of what is happening in the pipeline, not in log file. If you'd rather not print to user then we can remove this, and those that follow.

File: aasap (hunk 1 above)
Comment (slsevilla) [reply aasap:370]: removed path, added information to documentation, added fatal message after check

File: aasap (hunk 3 above)
Comment (slsevilla) [reply aasap:380]: done

File: aasap (hunk 7 above)
[the remark this answers is not in the document]
Comment (slsevilla) [reply aasap]: addressed in commit 870c2d98d82d931877b9cafcf73ac6ec58f53c35

File: aasap (hunk 8 above)
[the remark this answers is not in the document]
Comment (slsevilla) [reply aasap]: addressed in commit

File: aasap (hunk 1 above)
Comment (slsevilla) [reply aasap]: addressed in commits 870c2d98d82d931877b9cafcf73ac6ec58f53c35 1d919bf8293e9c4e1d0dbd402bfecce1e406e188

File: aasap (hunk 2 above)
Comment (slsevilla) [reply aasap:374]: addressed in commit 276842838f7dd3282fd40b6d354c298b0ee84068

File: aasap (hunk 3 above)
Comment (slsevilla) [reply aasap]: addressed in commit 9c599fb0a87a78455f6a39d75a0c36f838d67e18

**Comment by slsevilla** [comment]
> Also, please add `numpy` to the requirements.txt

added with commit b2d5c6d9823df61d586182820046fa6cbcb83b79

**Comment by slsevilla** [comment]
Hey @skchronicles ! I've complete the majority of tasks you've listed above in recent commits. The only one left to address is printing out updates to the user vs printing them to err(). Let me know your thoughts on this so I can handle this task (or leave it as is) or if there is anything I've missed.

Once this is merged, I'll work on adding in the parallelization and then changing everything from AASAP to METRO!

**Review by skchronicles — APPROVED** [review]
Okay, everything looks good. 

Thank you @slsevilla

**Pull request merged**